client/src/client.rs

```diff
@@ -443,44 +443,64 @@ impl Client {
         &self,
         transaction: Transaction,
     ) -> Result<HashOf<VersionedTransaction>> {
-        struct EventListenerInitialized;
-
         let client = self.clone();
-        let (event_sender, event_receiver) = mpsc::channel();
         let (init_sender, init_receiver) = mpsc::channel();
         let hash = transaction.hash();
-        let _handle = thread::spawn(move || -> eyre::Result<()> {
-            let event_iterator = client
-                .listen_for_events(PipelineEventFilter::new().hash(hash.into()).into())
-                .wrap_err("Failed to establish event listener connection.")?;
-            init_sender
-                .send(EventListenerInitialized)
-                .wrap_err("Failed to send init message through init channel.")?;
-            for event in event_iterator.flatten() {
-                if let Event::Pipeline(this_event) = event {
-                    match this_event.status {
-                        PipelineStatus::Validating => {}
-                        PipelineStatus::Rejected(reason) => event_sender
-                            .send(Err(reason))
-                            .wrap_err("Failed to send the transaction through event channel.")?,
-                        PipelineStatus::Committed => event_sender
-                            .send(Ok(hash.transmute()))
-                            .wrap_err("Failed to send the transaction through event channel.")?,
+
+        // TODO: use `std::thread::scope()` after update to Rust 1.63
+        let scope_res = crossbeam::scope(|spawner| {
+            let submitter_handle = spawner.spawn(move |_| -> Result<()> {
+                init_receiver
+                    .recv()
+                    .wrap_err("Failed to receive init message.")?;
+
+                self.submit_transaction(transaction)?;
+                Ok(())
+            });
+
+            let confirmation_res = client.listen_for_tx_confirmation(hash, &init_sender);
+
+            match submitter_handle.join() {
+                Ok(Ok(())) => confirmation_res,
+                Ok(Err(e)) => Err(e).wrap_err("Transaction submitter thread exited with error"),
+                Err(_) => Err(eyre!("Transaction submitter thread panicked")),
```

Comment:
Why discard the error? Why not include it in `wrap_err`.

Reply:
Same answer as above